designs/mixins.md

```diff
@@ -0,0 +1,1075 @@
+```
+
+```
+structure[id|name = ListSomethingInput] > structure
+```
+
+Both yield the following shapes:
+
+- `smithy.example#PaginatedInput`
+
+A `member` relationship exists from `ListSomethingInput` to the following
+shapes, in any order:
+
+- `smithy.example#ListSomethingInput$nameFilter`
+- `smithy.example#PaginatedInput$nextToken`
+- `smithy.example#PaginatedInput$pageSize`
+
+The following two selectors:
+
+```
+structure[id|name = ListSomethingInput] > member
+```
+
+```
+structure[id|name = ListSomethingInput] -[member]-> *
+```
+
+Both yield the following shapes, in any order:
+
+- `smithy.example#ListSomethingInput$nameFilter`
+- `smithy.example#PaginatedInput$nextToken`
+- `smithy.example#PaginatedInput$pageSize`
+
+
+### Mixins and member ordering
+
+The ordering of structure and union members is important for languages like C
+that require a stable ABI. Mixins provide a deterministic member ordering.
+Members are ordered by:
+
+- Members copied from mixins come before members defined directly in the shape.
+- Members are copied from mixins based on the order in which they applied to
+  a shape.
+
+Given the following model:
+
+```
+@mixin
+structure FilteredByName {
+    nameFilter: String
+}
+
+@mixin
+structure PaginatedInput {
+    nextToken: String,
+    pageSize: Integer
+}
+
+structure ListSomethingInput with PaginatedInput, FilteredByName {
+    sizeFilter: Integer
+}
+```
+
+The members are ordered as follows:
+
+- `nextToken`
+- `pageSize`
+- `nameFilter`
+- `sizeFilter`
+
+
+### Trait definitions
+
+#### `@mixin` trait
+
+The `@mixin` trait is a structured trait defined in the Smithy prelude as:
+
+```
+@trait(selector: ":is(structure, union)")
+structure mixin {
+    localTraits: LocalMixinTraitList
+}
+
+@private
+list LocalMixinTraitList {
+    member: LocalMixinTrait
+}
+
+@idRef(
+    selector: "[trait|trait]",
+    failWhenMissing: true,
+    errorMessage: """
+            Strings provided to the localTraits property of a mixin trait
+            must target a valid trait.""")
+@private
+string LocalMixinTrait
+```
+
+The `@mixin` trait has a single optional member named `localTraits` that
+contains a list of shape IDs. Each shape ID MUST reference a valid trait that
+is applied directly to the mixin. The traits referenced in `localTraits` are
+not copied onto shapes that use the mixin. `localTraits` only affects traits
+applied to the mixin container shape and has no impact on the members
+contained within a mixin.
+
+> Note: `smithy.api#mixin` is considered implicitly present in the
+> `localTraits` property and does not  need to be defined in the list of
+> local traits.
+
+
+#### `@override` trait
+
+The `@override` trait is an annotation trait defined in the Smithy prelude as:
+
+```
+@trait(selector: "is(structure, union) > member")
+structure override {}
+```
+
+
+### Reference implementation notes
+
+Mixins will be implemented in Smithy's Java reference implementation in a way
```

Comment:
Does this imply that other implementations could choose a different behavior, or is the reference implementation the preferred approach?

Reply:
This section is non-normative, and it's only specific to the reference implementation. Other implementations are free to implement however necessary. This document is not the specification, it's a design document. This is new to the repo, and it's the first one we've published. Once we have agreement here, we'll create a PR to add this to the specification (but also, we're not shipping this until we've finalized 1.1 of the IDL).